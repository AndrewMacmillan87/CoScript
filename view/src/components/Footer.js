import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./footer.css";

export const Footer = ({ user, loading, getMostRecentProject, ...props }) => {
    const authLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/profile'>Profile</Link>
            <Link
                to='#'
                onClick={async () => {
                    await getMostRecentProject();
                    const { projectData } = props;
                    projectData
                        ? props.history.push(`/editor${projectData.url}`)
                        : props.history.push("/editor");
                }}>
                Workspace
            </Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const guestLinks = (
        <div className='footer-links'>
            <Link to='/'>Home</Link>
            <Link to='/projects'>Active Projects</Link>
            <Link to='/publications'>Publications</Link>
        </div>
    );

    const isLoggedIn = user ? user.isLoggedIn : false;

    return (
        <footer className='footer'>
            <h2 className='footer-h2'>CoScript</h2>
            <div className='footer-container'>
                <div className='links-container socials'>
                    <p>Socials</p>
                    <div className='footer-links'>
                        <a href='https://github.com/AndrewMacmillan87/coscript'>
                            GitHub
                        </a>
                        <a href='/'>Twitter</a>
                        <a href='/'>Facebook</a>
                    </div>
                </div>
                <div className='links-container site-map'>
                    <p>Site Map</p>

                    {!loading && (
                        <Fragment>
                            {isLoggedIn ? authLinks : guestLinks}
                        </Fragment>
                    )}
                </div>
                <div className='links-container legals'>
                    <p>Legal</p>
                    <div className='footer-links'>
                        <a href='/terms'>Terms and Conditions</a>
                        <a href='/copyright'>Copyright Info</a>
                        <a href='/agreement'>Writers Agreements</a>
                        <a href='/'>Accreditation</a>
                    </div>
                </div>
            </div>
        </footer>
    );
};

Footer.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const mapProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(mapProps, { getMostRecentProject })(withRouter(Footer));
