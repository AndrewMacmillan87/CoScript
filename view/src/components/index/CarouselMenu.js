import React from "react";
import { withRouter } from "react-router-dom";
import ProjectCard from "../cards/ProjectCard";

import PropTypes from "prop-types";

export const CarouselMenu = ({
    projects,
    active,
    activeSegment,
    setActiveSegment,
    published,
    publishedSegment,
    setPublishedSegment,
    activeTab,
    activateTab,
    setActiveProjects,
    setPublishedProjects,
    loadProject,
    history,
}) => {
    const sky = require("../../img/sky-example.jpg");

    return (
        <section className='carousel'>
            <div className='carousel-container'>
                <div className='tabs-container'>
                    <div className={activeTab ? "tab" : "active tab"}>
                        <button
                            id='projects-tab'
                            className='tab-link'
                            onClick={() => {
                                activateTab();
                                setActiveProjects();
                            }}>
                            Projects
                        </button>
                    </div>
                    <div className={activeTab ? "active tab" : "tab"}>
                        <button
                            id='publications-tab'
                            className='tab-link'
                            onClick={() => {
                                activateTab();
                                setPublishedProjects();
                            }}>
                            Publications
                        </button>
                    </div>
                    <div className='carousel-line'></div>
                </div>
                <div className='carousel-items'>
                    <button
                        className='carousel-btn'
                        type='button'
                        onClick={() => {
                            if (active) {
                                if (activeSegment - 1 >= 0) {
                                    setActiveSegment(activeSegment - 1);
                                }
                            }
                            if (published) {
                                if (publishedSegment - 1 >= 0) {
                                    setPublishedSegment(publishedSegment - 1);
                                }
                            }
                        }}>
                        &lsaquo;
                    </button>
                    {projects && projects.length > 0 ? (
                        active ? (
                            projects[activeSegment].map((project) => {
                                return (
                                    <ProjectCard
                                        key={project.projectId}
                                        src={sky}
                                        img_alt='Project cover image'
                                        title={project.title}
                                        project_admin={project.author}
                                        genres={project.genres}
                                        onClick={async () => {
                                            await loadProject(project);
                                            history.push(
                                                `/landing${project.url}`
                                            );
                                        }}
                                    />
                                );
                            })
                        ) : (
                            published &&
                            projects[publishedSegment].map((project) => {
                                return (
                                    <ProjectCard
                                        key={project.projectId}
                                        src={sky}
                                        img_alt='Project cover image'
                                        title={project.title}
                                        project_admin={project.author}
                                        genres={project.genres}
                                        published={true}
                                        onClick={async () => {
                                            await loadProject(project);
                                            history.push(`/read${project.url}`);
                                        }}
                                    />
                                );
                            })
                        )
                    ) : (
                        <p>Projects coming soon...</p>
                    )}
                    <button
                        className='carousel-btn'
                        type='button'
                        onClick={() => {
                            if (active) {
                                if (activeSegment + 1 < projects.length) {
                                    setActiveSegment(activeSegment + 1);
                                }
                            }
                            if (published) {
                                if (publishedSegment + 1 < projects.length) {
                                    setPublishedSegment(publishedSegment + 1);
                                }
                            }
                        }}>
                        &rsaquo;
                    </button>
                </div>
            </div>
        </section>
    );
};

CarouselMenu.propTypes = {
    projects: PropTypes.array,
    active: PropTypes.bool,
    activeSegment: PropTypes.number,
    setActiveSegment: PropTypes.func,
    published: PropTypes.bool,
    publishedSegment: PropTypes.number,
    setPublishedSegment: PropTypes.func,
    activeTab: PropTypes.bool,
    activateTab: PropTypes.func,
    setActiveProjects: PropTypes.func,
    setPublishedProjects: PropTypes.func,
    loadProject: PropTypes.func,
    history: PropTypes.object,
};

export default withRouter(CarouselMenu);
