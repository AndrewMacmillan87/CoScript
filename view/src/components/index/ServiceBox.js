import React from "react";
import PropTypes from "prop-types";

const ServiceBox = ({ header, img, alt_text, description }) => {
    return (
        <div className='service-box'>
            <div>
                <h1>{header}</h1>
            </div>
            <div>
                <img className='service-icon' src={img} alt={alt_text} />
            </div>
            <div>
                <p>{description}</p>
            </div>
        </div>
    );
};

ServiceBox.propTypes = {
    header: PropTypes.string.isRequired,
    img: PropTypes.any,
    alt_text: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired
};

export default ServiceBox;
