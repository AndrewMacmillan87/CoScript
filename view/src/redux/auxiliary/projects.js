/**
 * Shuffles an array of projects randomly, in
 * place, to be displayed in the index page
 * carousel menu
 *
 * @param {Array} array - an array of projects
 */
export const shuffle = (array) => {
    if (array.length > 2) {
        for (let i = array.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [array[i], array[j]] = [array[j], array[i]];
        }
    }
};

/**
 * Breaks up a larger array of projects into a 2D
 * matrix of project array, each of length <width>.
 * Any remaining projects that equate to less than
 * <width> are added at the end of the 2D matrix
 *
 * @param {Array} array  - an array of projects
 * @param {Number} width - number of projects to show
 */
export const segment = (array, width) => {
    const segmentedArray = [];
    if (!array || !width) return segmentedArray;
    let tempArray = [];
    for (let i = 0; i < array.length; i++) {
        tempArray.push(array[i]);
        if (tempArray.length === width) {
            segmentedArray.push(tempArray);
            tempArray = [];
        }
    }
    if (tempArray.length > 0) segmentedArray.push(tempArray);
    return segmentedArray;
};

/**
 * Filters out projects belonging to this user from
 * an array of all projects
 *
 * @param {Array} array   - an array of projects
 * @param {String} userId - id of logged-in user
 */
export const filterUsersProjects = (array, userId) => {
    if (array.length > 0) {
        return array.filter((el) => el.user === userId, []);
    }
};

/**
 * Filters out all active projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterActive = (array) => {
    if (array.length > 0) {
        return array.filter((el) => !el.published, []);
    }
};

/**
 * Filters out all published projects from an array
 * of projects
 *
 * @param {Array} array - an array of projects
 */
export const filterPublished = (array) => {
    if (array.length > 0) {
        return array.filter((el) => el.published, []);
    }
};

/**
 * Loops over an array of projects to find any projects whose
 * subjects (author, title, genres) contain a matching sub-
 * string to the <searchCriteria>. Returns an array of projects
 * or an empty array if none were matched
 *
 * @param {Array} array           - an array of projects
 * @param {String} status         - the project's published status
 * @param {String} searchCriteria - the search term
 * @param {String} subject        - the field to search in
 */
export const filterByCriteria = (array, status, searchCriteria, subject) => {
    const cleanedSearch = searchCriteria.toLowerCase();
    const results = [];
    for (let project of array) {
        const condition =
            status === "active" ? !project.published : project.published;
        if (
            project[subject].toLowerCase().includes(cleanedSearch) &&
            condition
        ) {
            results.push(project);
        }
    }
    return results;
};
