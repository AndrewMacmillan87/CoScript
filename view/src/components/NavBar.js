import React, { Fragment } from "react";
import { Link, withRouter } from "react-router-dom";
import Button from "./buttons/Button";

import { getUserMessages } from "../redux/auxiliary/user-messages";

// Redux
import { connect } from "react-redux";
import { getMostRecentProject } from "../redux/actions/documents";

import PropTypes from "prop-types";

import "./navbar.css";

export class NavBar extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            visible: false,
            unread: false,
        };
        this.handleClick = this.handleClick.bind(this);
    }

    /**
     * Checks to see if the user has any unread messages
     * in their personal profile message inbox and sets
     * the 'unread' state property to true if they have.
     * This displays a small icon to the user to indicate
     * they have unread messages
     */
    async componentDidMount() {
        const userMessages = await getUserMessages();
        for (let message of userMessages) {
            if (!message.read) {
                this.setState({
                    unread: true,
                });
            }
        }
    }

    /**
     * Sets the 'visible' state property so that the
     * dropdown menu can be revealed and hidden by
     * clicking the link
     */
    handleClick() {
        this.setState((state) => ({
            visible: !state.visible,
        }));
    }

    render() {
        const { unread } = this.state;
        const { user, loading, history } = this.props;

        const menuIcon = require("../img/menu.svg");
        const spinner = require("../img/loading.svg");
        const isLoggedIn = user && user.isLoggedIn | false;

        const loggedIn = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                        <Link
                            to='#'
                            onClick={async () => {
                                await this.props.getMostRecentProject();
                                const { projectData } = this.props;
                                projectData
                                    ? history.push(`/editor${projectData.url}`)
                                    : history.push("/editor");
                            }}>
                            Workspace
                        </Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    {user && !loading ? (
                        <div className='logged-in-user-container'>
                            <p className='user-username'>
                                {user.profile.displayName}
                            </p>
                            <button
                                className='desktop-icon-link'
                                onClick={this.handleClick}>
                                {unread ? (
                                    <div className='avatar-unread-msg-container'>
                                        <div className='avatar-unread-msg-icon'></div>
                                        <img
                                            className='user-gravatar'
                                            src={user.avatar}
                                            alt='user gravatar'
                                        />
                                    </div>
                                ) : (
                                    <img
                                        className='user-gravatar'
                                        src={user.avatar}
                                        alt='user gravatar'
                                    />
                                )}
                            </button>
                        </div>
                    ) : (
                        <img
                            className='user-spinner'
                            src={spinner}
                            alt='loading spinner icon'
                        />
                    )}
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                    <Link
                        to='#'
                        onClick={async () => {
                            await this.props.getMostRecentProject();
                            const { projectData } = this.props;
                            projectData
                                ? history.push(`/editor${projectData.url}`)
                                : history.push("/editor");
                        }}>
                        <p className='dropdown-link'>Workspace</p>
                    </Link>
                    <Link to='/profile'>
                        <p className='dropdown-link'>My Profile</p>
                    </Link>
                    <a
                        onClick={() => {
                            history["from"] = "logout";
                            history.push("/redirect");
                        }}
                        href=' '>
                        <p className='dropdown-link'>Logout</p>
                    </a>
                </div>
            </Fragment>
        );

        const visiting = (
            <Fragment>
                <div className='header'>
                    <div className='logo-container'>
                        <Link to='/'>
                            <p className='logo'>CoScript</p>
                        </Link>
                    </div>
                    <div className='nav-links'>
                        <Link to='/projects'>Active Projects</Link>
                        <Link to='/publications'>Publications</Link>
                    </div>
                </div>
                <div className='nav-btns'>
                    <div>
                        <Link to='/register'>
                            <Button classes='btn' value='Register' />
                        </Link>
                    </div>
                    <div>
                        <Link to='/login'>
                            <Button classes='grn-btn' value='Login' />
                        </Link>
                    </div>
                </div>
                <button
                    className='dropdown-icon-link'
                    onClick={this.handleClick}>
                    <img
                        className='dropdown-icon'
                        src={menuIcon}
                        alt='Burger menu icon'
                    />
                </button>
                <div
                    id='dropdown-nav'
                    className={
                        this.state.visible
                            ? "dropdown-menu dropdown-content display"
                            : "dropdown-menu dropdown-content"
                    }>
                    <Link to='/register'>
                        <p className='dropdown-link'>Register</p>
                    </Link>
                    <Link to='/login'>
                        <p className='dropdown-link'>Login</p>
                    </Link>
                    <Link to='/projects'>
                        <p className='dropdown-link'>Active Projects</p>
                    </Link>
                    <Link to='/publications'>
                        <p className='dropdown-link'>Publications</p>
                    </Link>
                </div>
            </Fragment>
        );

        return (
            <nav className='navbar'>
                {!loading && (
                    <Fragment>{!isLoggedIn ? visiting : loggedIn}</Fragment>
                )}
            </nav>
        );
    }
}

NavBar.propTypes = {
    user: PropTypes.object,
    loading: PropTypes.bool,
    getMostRecentProject: PropTypes.func.isRequired,
};

const setProps = (state) => ({
    user: state.auth.user,
    loading: state.auth.loading,
    projectData: state.documents.projectData,
});

export default connect(setProps, {
    getMostRecentProject,
})(withRouter(NavBar));
