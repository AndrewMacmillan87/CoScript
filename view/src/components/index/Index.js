import React, { useState, useEffect } from "react";
import NavBar from "../NavBar";
import ServiceBox from "./ServiceBox";
import CarouselMenu from "./CarouselMenu";
import Footer from "../Footer";

import PropTypes from "prop-types";

import { stopSoundEffect } from "../../redux/auxiliary/typing-sound-effect";
import {
    shuffle,
    segment,
    filterActive,
    filterPublished,
} from "../../redux/auxiliary/projects";

// Redux
import { connect } from "react-redux";
import { clearFileData } from "../../redux/actions/files";
import { loadProject, loadAllProjects } from "../../redux/actions/documents";

import "./index.css";

export const Index = ({ ...props }) => {
    /**
     * Set the state for this functional component and its children
     */
    const [projects, setProjects] = useState(null);
    const [active, setActive] = useState(true);
    const [activeSegment, setActiveSegment] = useState(0);
    const [published, setPublished] = useState(false);
    const [publishedSegment, setPublishedSegment] = useState(0);
    const [activeTab, setActiveTab] = useState(false);

    /**
     * Destructure this components props
     */
    const { clearFileData, loadProject, loadAllProjects, projectArray } = props;

    /**
     * Scrolls back to the top of the page on mount and
     * unmount and loads all projects into the <projectArray>
     * prop. This effect contains an IIFE function to avoid
     * memory leaks in the event that the component unmounts
     * before the <loadAllProjects> async function has completed
     *
     */
    useEffect(() => {
        window.scrollTo(0, 0);
        (async function load() {
            await loadAllProjects();
        })();
    }, [loadAllProjects]);

    /**
     * Sets the browser tab title and scrolls to the
     * top of the page on mount. Also stops the typewriter
     * sound effect if the user navigated to this component
     * from an introductory workspace.
     *
     * A check is also performed on whether or not there
     * is any file data being persisted in Redux, if there
     * is, it is cleared so that the user can be shown the
     * introductory workspace if they have no personal
     * projects, or a recent personal project can be loaded
     * when they go back to the workspace
     */
    useEffect(() => {
        document.title = "CoScript - Writers Network";
        stopSoundEffect();
        clearFileData();
        const projects = projectArray ? projectArray : [];
        shuffle(projects);
        setProjects(
            setCarouselWidth(projects, window.innerWidth, filterActive)
        );
        window.addEventListener("resize", () => {
            setProjects(
                setCarouselWidth(projects, window.innerWidth, filterActive)
            );
        });
    }, [clearFileData, projectArray]);

    /**
     * Sets the carousel menu tabs 'Active' and
     * 'Published'
     */
    const makeActive = () => {
        setActiveTab(!activeTab);
    };

    /**
     * Checks the current width of the browser window
     * and sets the width of the carousel menu so it only
     * displays, 1, 2, 3 or 4 projects at a time. After
     * assigning a width, the <projectArray> is filtered
     * using the <filter> function to display either active
     * or published projects
     *
     * @param {Array} projectArray - an array of projects
     * @param {Number} screenWidth - current browser width
     * @param {Function} filter    - the filter function to use
     */
    function setCarouselWidth(projectArray, screenWidth, filter) {
        const projects = projectArray ? projectArray : [];
        if (projects.length > 0) {
            const filteredArray = filter(projects);
            let width;
            if (screenWidth <= 400) {
                width = 1;
            } else if (screenWidth <= 930) {
                width = 2;
            } else if (screenWidth <= 1100) {
                width = 3;
            } else {
                width = 4;
            }
            return segment(filteredArray, width);
        }
        return [];
    }

    /**
     * Loads all projects into the frontend when 'Active'
     * tab is clicked and shuffles them into a random order.
     * The active projects are then filtered out and displayed
     * to the user. The width of the carousel menu is then
     * set based on the 'window.innerWidth' property
     */
    async function setActiveProjects() {
        await loadAllProjects();
        const projects = projectArray ? projectArray : [];
        shuffle(projects);
        const carouselWidth = setCarouselWidth(
            projects,
            window.innerWidth,
            filterActive
        );
        setActive(true);
        setPublished(false);
        setProjects(carouselWidth);
    }

    /**
     * Loads all projects into the frontend when 'Published'
     * tab is clicked and shuffles them into a random order.
     * The published projects are then filtered out and displayed
     * to the user. The width of the carousel menu is then
     * set based on the 'window.innerWidth' property
     */
    async function setPublishedProjects() {
        await loadAllProjects();
        const projects = projectArray ? projectArray : [];
        shuffle(projects);
        const carouselWidth = setCarouselWidth(
            projects,
            window.innerWidth,
            filterPublished
        );
        setActive(false);
        setPublished(true);
        setProjects(carouselWidth);
    }

    const usersImg = require("../../img/users.svg");
    const collabImg = require("../../img/books.svg");
    const publishImg = require("../../img/archive.svg");

    return (
        <main>
            <NavBar />
            <section className='cover'>
                <div className='index-heading'>
                    <h1>CoScript</h1>
                    <p>Bringing writers together...</p>
                    <a className='down-arrow' href='#dummy'>
                        &darr;
                    </a>
                    <div id='dummy'></div>
                </div>
            </section>
            <section className='services'>
                <div className='services-container'>
                    <ServiceBox
                        header='-CONNECT-'
                        img={usersImg}
                        alt_text='Users icon'
                        description='Search for people and projects and start networking with others to get your project off the ground'
                    />
                    <ServiceBox
                        header='-COLLABORATE-'
                        img={collabImg}
                        alt_text='Book icon'
                        description='Create a workspace for your project with access to a rich text editor and start writing with your new found friends'
                    />
                    <ServiceBox
                        header='-PUBLISH-'
                        img={publishImg}
                        alt_text='Publish icon'
                        description='Upon completion, publish your final piece in the app to be read online and showcased for the world to see'
                    />
                </div>
            </section>
            <section className='home-page-about'>
                <div className='about-container'>
                    <div className='about-text-box'>
                        <h2 className='about-h2'>About CoScript</h2>
                        <div className='about-text'>
                            <p className='lead'>
                                This application aims to serve the writing
                                community, and anyone who is interested in
                                writing, as a meeting place and useful tool for
                                making collaboration on creative writing
                                projects easier than ever before. As a person
                                who has attempted to write a novel and several
                                short stories, but never actually finished any
                                of them, I can appreciate the demand for
                                collaborative spaces that offer help when you
                                need it.
                            </p>
                            <p>
                                The platform allows you to create your own
                                projects and put them out to the community. Once
                                your project is live, others can browse and
                                request to to get involved with the project, the
                                creator has total control over who can join.
                                Credit is given to ALL contributors whose
                                submissions were accepted by the project
                                creator, once a project has been published.
                            </p>
                            <p>
                                All active projects are available from the
                                'Active Projects' page, just browse and find
                                something that you're interested in and then
                                message the creator with a little intro to
                                request to be added to the project. If you've
                                got an idea and are willing to make it a
                                collaborative project then sign up today and
                                start you're journey towards being a published
                                author!
                            </p>
                        </div>
                    </div>
                </div>
            </section>
            <CarouselMenu
                projects={projects}
                active={active}
                activeSegment={activeSegment}
                setActiveSegment={setActiveSegment}
                published={published}
                publishedSegment={publishedSegment}
                setPublishedSegment={setPublishedSegment}
                activeTab={activeTab}
                activateTab={makeActive}
                setActiveProjects={setActiveProjects}
                setPublishedProjects={setPublishedProjects}
                loadProject={loadProject}
            />
            <Footer />
        </main>
    );
};

Index.propTypes = {
    clearFileData: PropTypes.func,
    loadProject: PropTypes.func,
    loadAllProjects: PropTypes.func,
};

const mapProps = (state) => ({
    projectArray: state.documents.projectArray,
});

export default connect(mapProps, {
    clearFileData,
    loadProject,
    loadAllProjects,
})(Index);
