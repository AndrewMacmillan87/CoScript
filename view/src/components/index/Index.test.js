import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import { mount } from "enzyme";

import { Provider } from "react-redux";
import { reduxStore } from "../../redux/store/reduxStore";

import { Index } from "./Index";
import NavBar from "../NavBar";
import ServiceBox from "./ServiceBox";
import { CarouselMenu } from "./CarouselMenu";
import Footer from "../Footer";

import * as projectsActions from "../../redux/auxiliary/projects";

const originalShuffle = projectsActions.shuffle;

describe("Index component", () => {
    let mockClearFileData, mockLoadAllProjects, props;
    beforeEach(() => {
        global.HTMLMediaElement.prototype.pause = jest.fn();
        global.scrollTo = jest.fn();
        mockClearFileData = jest.fn();
        mockLoadAllProjects = jest.fn();
        props = {
            clearFileData: mockClearFileData,
            loadAllProjects: mockLoadAllProjects,
        };
    });

    afterAll(() => {
        projectActions.shuffle = originalShuffle;
    });

    test("renders without crashing", () => {
        const div = document.createElement("div");
        ReactDOM.render(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>,
            div
        );
        ReactDOM.unmountComponentAtNode(div);
    });

    it("renders correctly", () => {
        const tree = renderer
            .create(
                <Provider store={reduxStore}>
                    <Router>
                        <Index {...props} />
                    </Router>
                </Provider>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render a main element", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("main").length).toEqual(1);
    });

    it("should render the NavBar component on mount", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(NavBar).length).toEqual(1);
    });

    it("should render 4 section elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("section").length).toEqual(4);
    });

    it("should render 3 ServiceBox components on mount", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(ServiceBox).length).toEqual(3);
    });

    it("should render 11 p tags", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("p").length).toEqual(11);
    });

    it("should render the CarouselMenu component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(CarouselMenu).length).toEqual(1);
    });

    it("should render the Footer component", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find(Footer).length).toEqual(1);
    });

    it("should call `clearFileData()` on mount", () => {
        mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(mockClearFileData).toBeCalled();
    });

    it("should call `mockLoadAllProjects()` on mount", () => {
        mount(
            <Provider store={reduxStore}>
                <Router>
                    <Index {...props} />
                </Router>
            </Provider>
        );
        expect(mockLoadAllProjects).toBeCalled();
    });
});

describe("CarouselMenu component", () => {
    let mockLoadAllProjects, mockLoadProject, props;
    beforeEach(() => {
        global.HTMLMediaElement.prototype.pause = jest.fn();
        global.scrollTo = jest.fn();
        mockLoadProject = jest.fn(() => Promise.resolve());
        mockLoadAllProjects = jest.fn(() => Promise.resolve());
        props = {
            loadProject: mockLoadProject,
            activeTab: true,
            onClick: jest.fn(),
        };
    });

    afterAll(() => {
        projectsActions.shuffle = originalShuffle;
    });

    it("should render a section element with a className of carousel", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <CarouselMenu {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find({ className: "carousel" }).length).toEqual(1);
    });

    it("should render 4 button elements", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <CarouselMenu {...props} />
                </Router>
            </Provider>
        );
        expect(wrapper.find("button").length).toEqual(4);
    });

    it("should render a div element with a className of carousel-container", () => {
        const wrapper = mount(
            <Provider store={reduxStore}>
                <Router>
                    <CarouselMenu {...props} />
                </Router>
            </Provider>
        );
        expect(
            wrapper.find({ className: "carousel-container" }).length
        ).toEqual(1);
    });
});
