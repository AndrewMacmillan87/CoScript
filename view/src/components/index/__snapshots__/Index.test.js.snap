// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index component renders correctly 1`] = `
<main>
  <nav
    className="navbar"
  />
  <section
    className="cover"
  >
    <div
      className="index-heading"
    >
      <h1>
        CoScript
      </h1>
      <p>
        Bringing writers together...
      </p>
      <a
        className="down-arrow"
        href="#dummy"
      >
        ↓
      </a>
      <div
        id="dummy"
      />
    </div>
  </section>
  <section
    className="services"
  >
    <div
      className="services-container"
    >
      <div
        className="service-box"
      >
        <div>
          <h1>
            -CONNECT-
          </h1>
        </div>
        <div>
          <img
            alt="Users icon"
            className="service-icon"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "users.svg",
              }
            }
          />
        </div>
        <div>
          <p>
            Search for people and projects and start networking with others to get your project off the ground
          </p>
        </div>
      </div>
      <div
        className="service-box"
      >
        <div>
          <h1>
            -COLLABORATE-
          </h1>
        </div>
        <div>
          <img
            alt="Book icon"
            className="service-icon"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "books.svg",
              }
            }
          />
        </div>
        <div>
          <p>
            Create a workspace for your project with access to a rich text editor and start writing with your new found friends
          </p>
        </div>
      </div>
      <div
        className="service-box"
      >
        <div>
          <h1>
            -PUBLISH-
          </h1>
        </div>
        <div>
          <img
            alt="Publish icon"
            className="service-icon"
            src={
              Object {
                "ReactComponent": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                },
                "__esModule": true,
                "default": "archive.svg",
              }
            }
          />
        </div>
        <div>
          <p>
            Upon completion, publish your final piece in the app to be read online and showcased for the world to see
          </p>
        </div>
      </div>
    </div>
  </section>
  <section
    className="home-page-about"
  >
    <div
      className="about-container"
    >
      <div
        className="about-text-box"
      >
        <h2
          className="about-h2"
        >
          About CoScript
        </h2>
        <div
          className="about-text"
        >
          <p
            className="lead"
          >
            This application aims to serve the writing community, and anyone who is interested in writing, as a meeting place and useful tool for making collaboration on creative writing projects easier than ever before. As a person who has attempted to write a novel and several short stories, but never actually finished any of them, I can appreciate the demand for collaborative spaces that offer help when you need it.
          </p>
          <p>
            The platform allows you to create your own projects and put them out to the community. Once your project is live, others can browse and request to to get involved with the project, the creator has total control over who can join. Credit is given to ALL contributors whose submissions were accepted by the project creator, once a project has been published.
          </p>
          <p>
            All active projects are available from the 'Active Projects' page, just browse and find something that you're interested in and then message the creator with a little intro to request to be added to the project. If you've got an idea and are willing to make it a collaborative project then sign up today and start you're journey towards being a published author!
          </p>
        </div>
      </div>
    </div>
  </section>
  <section
    className="carousel"
  >
    <div
      className="carousel-container"
    >
      <div
        className="tabs-container"
      >
        <div
          className="active tab"
        >
          <button
            className="tab-link"
            id="projects-tab"
            onClick={[Function]}
          >
            Projects
          </button>
        </div>
        <div
          className="tab"
        >
          <button
            className="tab-link"
            id="publications-tab"
            onClick={[Function]}
          >
            Publications
          </button>
        </div>
        <div
          className="carousel-line"
        />
      </div>
      <div
        className="carousel-items"
      >
        <button
          className="carousel-btn"
          onClick={[Function]}
          type="button"
        >
          ‹
        </button>
        <p>
          Projects coming soon...
        </p>
        <button
          className="carousel-btn"
          onClick={[Function]}
          type="button"
        >
          ›
        </button>
      </div>
    </div>
  </section>
  <footer
    className="footer"
  >
    <h2
      className="footer-h2"
    >
      CoScript
    </h2>
    <div
      className="footer-container"
    >
      <div
        className="links-container socials"
      >
        <p>
          Socials
        </p>
        <div
          className="footer-links"
        >
          <a
            href="https://github.com/AndrewMacmillan87/coscript"
          >
            GitHub
          </a>
          <a
            href="/"
          >
            Twitter
          </a>
          <a
            href="/"
          >
            Facebook
          </a>
        </div>
      </div>
      <div
        className="links-container site-map"
      >
        <p>
          Site Map
        </p>
      </div>
      <div
        className="links-container legals"
      >
        <p>
          Legal
        </p>
        <div
          className="footer-links"
        >
          <a
            href="/terms"
          >
            Terms and Conditions
          </a>
          <a
            href="/copyright"
          >
            Copyright Info
          </a>
          <a
            href="/agreement"
          >
            Writers Agreements
          </a>
          <a
            href="/"
          >
            Accreditation
          </a>
        </div>
      </div>
    </div>
  </footer>
</main>
`;
